File: eddy-audio-main/bindings/csharp/WhisperPipeline.cs
// Copyright (C) 2025 Fluid Inference
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Eddy;

/// <summary>
/// Configuration for Whisper pipeline
/// </summary>
public class WhisperConfig
{
    /// <summary>
    /// Path to the Whisper model directory (containing .xml/.bin files)
    /// </summary>
    public required string ModelPath { get; set; }

    /// <summary>
    /// Device to run inference on: "NPU", "CPU", or "AUTO"
    /// </summary>
    public string Device { get; set; } = "NPU";

    /// <summary>
    /// Language code (e.g., "en", "zh", "es") or "auto" for auto-detection
    /// </summary>
    public string Language { get; set; } = "en";

    /// <summary>
    /// Task: "transcribe" or "translate" (translate to English)
    /// </summary>
    public string Task { get; set; } = "transcribe";

    /// <summary>
    /// Whether to return word-level timestamps
    /// </summary>
    public bool ReturnTimestamps { get; set; } = true;

    /// <summary>
    /// Enable model compilation caching (recommended for NPU)
    /// </summary>
    public bool EnableCache { get; set; } = true;

    /// <summary>
    /// Cache directory for compiled models (null = default OpenVINO cache location)
    /// </summary>
    public string? CacheDir { get; set; } = null;
}

/// <summary>
/// A chunk of transcribed text with timestamps
/// </summary>
public class WhisperChunk
{
    /// <summary>
    /// Start time in seconds
    /// </summary>
    public float StartTime { get; init; }

    /// <summary>
    /// End time in seconds (-1.0 if not available)
    /// </summary>
    public float EndTime { get; init; }

    /// <summary>
    /// Transcribed text for this chunk
    /// </summary>
    public required string Text { get; init; }
}

/// <summary>
/// Result from Whisper transcription
/// </summary>
public class WhisperResult
{
    /// <summary>
    /// Full transcribed text
    /// </summary>
    public required string Text { get; init; }

    /// <summary>
    /// Word/segment-level chunks with timestamps (if enabled)
    /// </summary>
    public List<WhisperChunk> Chunks { get; init; } = new();

    /// <summary>
    /// Average confidence score (0.0 to 1.0)
    /// </summary>
    public float Confidence { get; init; }

    /// <summary>
    /// Inference duration in milliseconds
    /// </summary>
    public double InferenceDurationMs { get; init; }
}

/// <summary>
/// Whisper speech recognition pipeline
///
/// Wraps OpenVINO GenAI's WhisperPipeline for easy audio transcription.
/// Supports language selection, timestamps, and NPU acceleration.
/// </summary>
public class WhisperPipeline : IDisposable
{
    private IntPtr _handle;
    private bool _disposed = false;

    /// <summary>
    /// Construct a WhisperPipeline with the given configuration
    ///
    /// Note: First run with NPU device may take 5+ minutes for model compilation.
    /// Subsequent runs will be fast if caching is enabled.
    /// </summary>
    /// <param name="config">Configuration parameters</param>
    /// <exception cref="EddyException">Thrown if model loading fails</exception>
    public WhisperPipeline(WhisperConfig config)
    {
        var modelPathPtr = Marshal.StringToHGlobalAnsi(config.ModelPath);
        var devicePtr = Marshal.StringToHGlobalAnsi(config.Device);
        var languagePtr = Marshal.StringToHGlobalAnsi(config.Language);
        var taskPtr = Marshal.StringToHGlobalAnsi(config.Task);
        var cacheDirPtr = config.CacheDir != null
            ? Marshal.StringToHGlobalAnsi(config.CacheDir)
            : IntPtr.Zero;

        try
        {
            var nativeConfig = new Native.EddyWhisperConfig
            {
                model_path = modelPathPtr,
                device = devicePtr,
                language = languagePtr,
                task = taskPtr,
                return_timestamps = config.ReturnTimestamps,
                enable_cache = config.EnableCache,
                cache_dir = cacheDirPtr
            };

            _handle = Native.eddy_whisper_create(nativeConfig, out var errorMsg);

            if (_handle == IntPtr.Zero)
            {
                var error = Native.PtrToStringAndFree(errorMsg) ?? "Unknown error";
                throw new EddyException($"Failed to create Whisper pipeline: {error}");
            }
        }
        finally
        {
            Marshal.FreeHGlobal(modelPathPtr);
            Marshal.FreeHGlobal(devicePtr);
            Marshal.FreeHGlobal(languagePtr);
            Marshal.FreeHGlobal(taskPtr);
            if (cacheDirPtr != IntPtr.Zero)
                Marshal.FreeHGlobal(cacheDirPtr);
        }
    }

    /// <summary>
    /// Transcribe audio from a WAV file
    /// </summary>
    /// <param name="wavPath">Path to WAV file (must be 16kHz, mono or stereo)</param>
    /// <returns>WhisperResult containing transcribed text and metadata</returns>
    /// <exception cref="EddyException">Thrown if transcription fails</exception>
    public WhisperResult Transcribe(string wavPath)
    {
        ThrowIfDisposed();

        var error = Native.eddy_whisper_transcribe_file(
            _handle,
            wavPath,
            out var result,
            out var errorMsg
        );

        if (error != Native.EddyError.EDDY_OK)
        {
            var errStr = Native.PtrToStringAndFree(errorMsg) ?? "Unknown error";
            throw new EddyException($"Transcription failed: {errStr}");
        }

        try
        {
            return ConvertResult(result);
        }
        finally
        {
            Native.eddy_whisper_free_result(ref result);
        }
    }

    /// <summary>
    /// Transcribe audio from raw PCM float32 buffer
    /// </summary>
    /// <param name="pcm">Float32 PCM samples (normalized to [-1, 1])</param>
    /// <param name="sampleRate">Sample rate in Hz (default 16000)</param>
    /// <returns>WhisperResult containing transcribed text and metadata</returns>
    /// <exception cref="EddyException">Thrown if transcription fails</exception>
    public WhisperResult Transcribe(float[] pcm, int sampleRate = 16000)
    {
        ThrowIfDisposed();

        var error = Native.eddy_whisper_transcribe_buffer(
            _handle,
            pcm,
            (nuint)pcm.Length,
            sampleRate,
            out var result,
            out var errorMsg
        );

        if (error != Native.EddyError.EDDY_OK)
        {
            var errStr = Native.PtrToStringAndFree(errorMsg) ?? "Unknown error";
            throw new EddyException($"Transcription failed: {errStr}");
        }

        try
        {
            return ConvertResult(result);
        }
        finally
        {
            Native.eddy_whisper_free_result(ref result);
        }
    }

    /// <summary>
    /// Set the language for transcription
    /// </summary>
    /// <param name="language">Language code (e.g., "en", "zh") or "auto"</param>
    public void SetLanguage(string language)
    {
        ThrowIfDisposed();
        Native.eddy_whisper_set_language(_handle, language);
    }

    /// <summary>
    /// Set the task (transcribe or translate)
    /// </summary>
    /// <param name="task">"transcribe" or "translate"</param>
    public void SetTask(string task)
    {
        ThrowIfDisposed();
        Native.eddy_whisper_set_task(_handle, task);
    }

    /// <summary>
    /// Get the current language setting
    /// </summary>
    public string GetLanguage()
    {
        ThrowIfDisposed();
        var ptr = Native.eddy_whisper_get_language(_handle);
        return Marshal.PtrToStringAnsi(ptr) ?? "en";
    }

    private WhisperResult ConvertResult(Native.EddyWhisperResult nativeResult)
    {
        var text = Marshal.PtrToStringAnsi(nativeResult.text) ?? "";
        var chunks = new List<WhisperChunk>();

        if (nativeResult.num_chunks > 0 && nativeResult.chunks != IntPtr.Zero)
        {
            var chunkSize = Marshal.SizeOf<Native.EddyWhisperChunk>();
            for (nuint i = 0; i < nativeResult.num_chunks; i++)
            {
                var chunkPtr = IntPtr.Add(nativeResult.chunks, (int)(i * (nuint)chunkSize));
                var nativeChunk = Marshal.PtrToStructure<Native.EddyWhisperChunk>(chunkPtr);
                var chunkText = Marshal.PtrToStringAnsi(nativeChunk.text) ?? "";

                chunks.Add(new WhisperChunk
                {
                    StartTime = nativeChunk.start_ts,
                    EndTime = nativeChunk.end_ts,
                    Text = chunkText
                });
            }
        }

        return new WhisperResult
        {
            Text = text,
            Chunks = chunks,
            Confidence = nativeResult.confidence,
            InferenceDurationMs = nativeResult.inference_duration_ms
        };
    }

    private void ThrowIfDisposed()
    {
        if (_disposed)
            throw new ObjectDisposedException(nameof(WhisperPipeline));
    }

    public void Dispose()
    {
        if (_disposed) return;

        if (_handle != IntPtr.Zero)
        {
            Native.eddy_whisper_destroy(_handle);
            _handle = IntPtr.Zero;
        }

        _disposed = true;
        GC.SuppressFinalize(this);
    }

    ~WhisperPipeline()
    {
        Dispose();
    }
}

/// <summary>
/// Exception thrown by eddy
/// </summary>
public class EddyException : Exception
{
    public EddyException(string message) : base(message) { }
    public EddyException(string message, Exception inner) : base(message, inner) { }
}

File: eddy-audio-main/src/pipelines/whisper_pipeline.cpp
// Copyright (C) 2025 Eddy SDK
// SPDX-License-Identifier: Apache-2.0

#include "eddy/pipelines/whisper_pipeline.hpp"
#include "eddy/utils/audio_utils.hpp"

#include <openvino/genai/whisper_pipeline.hpp>
#include <openvino/openvino.hpp>

#include <chrono>
#include <iostream>
#include <stdexcept>

namespace eddy {

struct WhisperPipeline::Impl {
    WhisperConfig config;
    std::unique_ptr<ov::genai::WhisperPipeline> pipeline;

    explicit Impl(const WhisperConfig& cfg) : config(cfg) {
        std::cout << "[Eddy] Initializing Whisper pipeline..." << std::endl;
        std::cout << "[Eddy] Model: " << config.model_path << std::endl;
        std::cout << "[Eddy] Device: " << config.device << std::endl;

        // Set up OpenVINO properties
        ov::AnyMap properties;

        // Enable model caching if requested
        if (config.enable_cache && !config.cache_dir.empty()) {
            properties[ov::cache_dir.name()] = config.cache_dir;
            std::cout << "[Eddy] Cache directory: " << config.cache_dir << std::endl;
        }

        std::cout << "[Eddy] Loading model... (First run on NPU may take 5+ minutes for compilation)" << std::endl;
        auto start = std::chrono::high_resolution_clock::now();

        // Create the OpenVINO GenAI WhisperPipeline
        pipeline = std::make_unique<ov::genai::WhisperPipeline>(
            config.model_path,
            config.device,
            properties
        );

        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
        std::cout << "[Eddy] Model loaded in " << duration.count() << " ms" << std::endl;

        // Configure generation settings
        auto gen_config = pipeline->get_generation_config();

        // Set language
        if (config.language != "auto") {
            gen_config.language = "<|" + config.language + "|>";
        }

        // Set task
        gen_config.task = config.task;

        // Enable timestamps
        gen_config.return_timestamps = config.return_timestamps;

        pipeline->set_generation_config(gen_config);

        std::cout << "[Eddy] Whisper pipeline ready!" << std::endl;
    }
};

WhisperPipeline::WhisperPipeline(const WhisperConfig& config)
    : impl_(std::make_unique<Impl>(config)) {}

WhisperPipeline::~WhisperPipeline() = default;

WhisperPipeline::WhisperPipeline(WhisperPipeline&&) noexcept = default;
WhisperPipeline& WhisperPipeline::operator=(WhisperPipeline&&) noexcept = default;

WhisperResult WhisperPipeline::transcribe(const std::string& wav_path) {
    std::cout << "[Eddy] Reading audio file: " << wav_path << std::endl;

    // Read WAV file
    auto pcm = audio::read_wav(wav_path);

    std::cout << "[Eddy] Audio loaded: " << pcm.size() << " samples ("
              << (pcm.size() / 16000.0) << " seconds)" << std::endl;

    // Transcribe
    return transcribe(pcm.data(), pcm.size(), 16000);
}

WhisperResult WhisperPipeline::transcribe(const float* pcm, size_t length, int sample_rate) {
    if (sample_rate != 16000) {
        throw std::runtime_error("Only 16kHz sample rate is supported, got " + std::to_string(sample_rate));
    }

    std::cout << "[Eddy] Running transcription..." << std::endl;
    auto start = std::chrono::high_resolution_clock::now();

    // Create input
    ov::genai::RawSpeechInput input(pcm, pcm + length);

    // Run inference
    auto ov_result = impl_->pipeline->generate(input);

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);

    // Convert result
    WhisperResult result;

    if (!ov_result.texts.empty()) {
        result.text = ov_result.texts[0];
    }

    if (!ov_result.scores.empty()) {
        result.confidence = ov_result.scores[0];
    }

    // Convert chunks if available
    if (ov_result.chunks.has_value()) {
        for (const auto& chunk : *ov_result.chunks) {
            WhisperChunk eddy_chunk;
            eddy_chunk.start_ts = chunk.start_ts;
            eddy_chunk.end_ts = chunk.end_ts;
            eddy_chunk.text = chunk.text;
            result.chunks.push_back(eddy_chunk);
        }
    }

    result.inference_duration_ms = static_cast<double>(duration.count());

    std::cout << "[Eddy] Transcription complete in " << result.inference_duration_ms << " ms" << std::endl;
    std::cout << "[Eddy] Result: " << result.text << std::endl;

    return result;
}

void WhisperPipeline::set_language(const std::string& language) {
    impl_->config.language = language;

    auto gen_config = impl_->pipeline->get_generation_config();
    if (language != "auto") {
        gen_config.language = "<|" + language + "|>";
    } else {
        gen_config.language = std::nullopt;
    }
    impl_->pipeline->set_generation_config(gen_config);

    std::cout << "[Eddy] Language set to: " << language << std::endl;
}

std::string WhisperPipeline::get_language() const {
    return impl_->config.language;
}

void WhisperPipeline::set_task(const std::string& task) {
    impl_->config.task = task;

    auto gen_config = impl_->pipeline->get_generation_config();
    gen_config.task = task;
    impl_->pipeline->set_generation_config(gen_config);

    std::cout << "[Eddy] Task set to: " << task << std::endl;
}

const WhisperConfig& WhisperPipeline::get_config() const {
    return impl_->config;
}

}  // namespace eddy